### ISSUE TITLE
adding BatchNormalization layer yielded a very large value

### ISSUE THREAD

**Issue by santisy** [torch/nn#770]
Hi! 
the BatchNormalization layer really caused me trouble. I was writing a regular CNN and test it with all
input entry equals zero, however, I got either -1 or 1, which means that the value had been so large before the network reached layer Tanh.

Here is my codes:
```
netWorks.buildNetU = function()
  local picDim = netWorks.picDim
  local actionNum = netWorks.actionNum

  local uModel = nn.Sequential()

  uModel:add(nn.CustomSpConv(3, 32, 5, 5))
  uModel:add(nn.SpatialBatchNormalization(4))
  uModel:add(nn.ReLU())
  uModel:add(nn.CustomSpConv(32, 32, 5, 5))
  uModel:add(nn.SpatialBatchNormalization(4))
  uModel:add(nn.ReLU())
  uModel:add(nn.CustomSpConv(32, 32, 5, 5))
  uModel:add(nn.SpatialBatchNormalization(4))
  uModel:add(nn.ReLU())

  local convResultsNum = (picDim - 5*3 + 1*3)^2*32

  uModel:add(nn.Reshape(convResultsNum))

  uModel:add(nn.CustomLinear(convResultsNum, 200))
  uModel:add(nn.BatchNormalization(2))
  uModel:add(nn.ReLU())
  uModel:add(nn.CustomLinear(200, 200))
  uModel:add(nn.BatchNormalization(2))
  uModel:add(nn.ReLU())

  local finalLayer = nn.Linear(200, actionNum)
  finalLayer.weight = torch.Tensor(actionNum, 200):uniform(-3e-3, 3e-3)
  finalLayer.bias = torch.Tensor(actionNum):uniform(-3e-3, 3e-3)

  uModel:add(finalLayer)
  uModel:add(nn.Tanh())

  -- make it use CUDA!!
  uModel:cuda()

  return uModel
end
```

And the `CustomLinear` and `CustomSpConv` are just two wrappers of special bias and weights
initialization. Here are their codes:
```
nn.CustomSpConv = function (nInputPlane, nOutputPlane, kW, kH)
  local conv_layer = nn.SpatialConvolution(nInputPlane, nOutputPlane, kW, kH)
  local bounds = torch.sqrt(nOutputPlane*kW*kH)
  conv_layer.weight = torch.Tensor(nOutputPlane, nInputPlane, kH, kW):uniform(-1/bounds, 1/bounds)
  conv_layer.bias = torch.Tensor(nOutputPlane):uniform(-1/bounds, 1/bounds)

  return conv_layer
end
nn.CustomLinear = function (inputDim, outputDim)
  local linear_layer = nn.Linear(inputDim, outputDim)
  local bounds = torch.sqrt(outputDim)
  linear_layer.weight = torch.Tensor(outputDim, inputDim):uniform(-1/bounds,1/bounds)
  linear_layer.bias = torch.Tensor(outputDim):uniform(-1/bounds, 1/bounds)
  return linear_layer
end
```
the parameters: `picDim = 60, actionNum = 3`
And I try `print(uModel:forward(torch.zeros(16,3,60,60):cuda()))`. Got all -1

And when I removed all the BatchNormalization layers(including the spatial ones), I got the desired value.(the value near 0)

**Comment by fmassa** [comment]
This is expected. If the input is all zeros, after the first convolution, you will have a constant value per feature map (the bias of the convolution), thus in the batch norm as the mean per feature map is the bias and the std is zero, so `(x-mean)/std` is quite unstable  (`nan` if no care is taken). There is an `eps` element in order to avoid `nan`s. Anyway, after batch norm all elements per feature map will still be the same, and the process continues.
Also, this discussion is best suited for the Google Groups https://groups.google.com/forum/#!forum/torch7

**Comment by santisy** [comment]
my faults, I've thought it out.
